File: WindowsController/UiApi/UIApi.cs
using System.Linq;
using System.Threading.Tasks;
using LiveObjects.ModelDescription;
using Newtonsoft.Json;

namespace ApiHooker.UiApi
{
    public class UIApi: ILiveObject
    {
        public string ResourceId => "api";

        [Publish]
        public async Task<string> EchoAsync(string message) => $"echo response: {message}";

        [Publish]
        public async Task<UIProcess> LaunchAndInjectAsync(string path)
        {
            return new UIProcess { Path = path };
        }

        [Publish]
        public async Task<UIHookableMethod[]> GetHookableMethodsAsync()
        {
            return typeof(ApiDefinitions).GetFields().Select(f => new UIHookableMethod { Name = f.Name }).ToArray();
        }
    }
}

File: ApiHookerUI/src/RemoteModel/UIApi.ts
import { JsonRpc, UIObject } from '../JsonRpc';
import UIHookableMethod from './UIHookableMethod';
import UIProcess from "./UIHookedProcess";

export default class UIApi extends UIObject {
    public hookableMethods = UIObject.typedArray(UIHookableMethod);
    public hookedProcesses = UIObject.typedArray(UIProcess);

    echo(message: string): Promise<string> {
        return this.jsonRpc.call(this.resourceId, "Echo", [message]).then(x => <string>x);
    }

    launchAndHook(path: string): Promise<UIProcess> {
        return this.jsonRpc.call(this.resourceId, "LaunchAndHook", [path])
            .then(x => this.jsonRpc.createFrom(UIProcess, x));
    }
};